### DISCUSSION TITLE
Correct version of kube

### DISCUSSION THREAD

**Discussion by pisees** [kube-rs/kube#434] [Q&A] [answered]
Regarding Installation and "Select a version of kube along with the generated k8s-openapi types corresponding for your cluster version:"

How do I find out the correct versions in my cluster?

**Comment by kazk** [comment, marked as answer]
`kubectl version --short` should show versions of your client and server. Something like

```
Client Vesion: v1.19.4
Server Version: v1.20.2
```

Use the version matching your server version.

**Reply by DevasiaThomas** [reply]
Is it possible to compile a client that can interact with multiple server versions?? Or can you only target one server version?

**Reply by nightkr** [reply]
The version that you select dictates the types and fields that you can see. *In general* it should be safe to target the oldest version that you support.

The exception is that some Kubernetes versions remove old alpha/beta APIs, in which case any request for those APIs will fail at runtime. For example, K8s 1.22 removed support for CustomResourceDefinition v1beta1.

If you want to take (optional) advantage of a feature from a newer version then you will need to target (at least) that version instead, but then it's on you to to fall back when those features are not available in older versions (that you still support).

**Reply by DevasiaThomas** [reply]
> If you want to take (optional) advantage of a feature from a newer version then you will need to target (at least) that version instead, but then it's on you to to fall back when those features are not available in older versions (that you still support).

Thanks for the quick reply @teozkr . When you say the above, are you saying can target multiple in the same compile?  Right now I'm just interested in making watchers for Pod events

**Reply by nightkr** [reply]
No, you can only enable one version feature at a time. If you enable the newer version then you need to disable the older one.

**Reply by DevasiaThomas** [reply]
So what did you mean by "its on me to fall back when it's not available"?  Can this be done at runtime instead of compile time?

**Reply by nightkr** [reply]
>  Can this be done at runtime instead of compile time?

Yes.

For example: `CronJob` v1 was added in K8s v1.21. If you want to be compatible with K8s v1.20 then you have two different choices:

1. Always write `CronJob` v1beta1. This means that you can't take advantage of features that are only exposed in v1, even if the K8s cluster you're running against has it available.
2. Try to write a v1 object, but fall back to v1beta1 if that doesn't exist. For example:

```rust
let kube = kube::Client::try_default().await?;
let cronjobs_v1 = kube::Api::<k8s_openapi::api::batch::v1::CronJob>::default_namespaced(kube.clone());
let cronjobs_v1b1 = kube::Api::<k8s_openapi::api::batch::v1beta1::CronJob>::default_namespaced(kube);
match cronjobs_v1.create(&PostParams::default(), build_cj_v1()).await {
  Ok(_) => {}
  Err(_) => {
    cronjobs_v1b1.create(&PostParams::default(), build_cj_v1b1()).await?;
  }
}
```

In general I'd recommend approach 1 unless you have a good reason to avoid it.. it's easy to end up with subtle desyncs otherwise.

**Reply by clux** [reply]
Yeah, if you want type safety you gotta get it from somewhere. I wouldn't try to use solution 2 there myself because it's generally super easy to do an upgrade manually:

```diff
-use k8s_openapi::api::batch::v1beta::CronJob
+use k8s_openapi::api::batch::v1::CronJob
```

if you want to dynamically figure out what CronJob is available in the cluster and use that instead, then you can use the `discovery` api with say a [`pinned_group`](https://docs.rs/kube/0.63.2/kube/discovery/oneshot/fn.pinned_group.html). But note that you don't get typesafety of the struct you are serializing into anymore then, because you don't actually know what struct you are getting.

**Reply by DevasiaThomas** [reply]
@teozkr Cool! Thanks for the example and the recommendation . A follow up question would be in case of solution 2, what would I specify in the ``Cargo.toml`` file against ``features`` for ``k8s-openapi``?

**Reply by nightkr** [reply]
@DevasiaThomas You would generally want to specify the oldest K8s version that contains both versions. In this case, that would be `v1_21` (since `CronJob` v1 was introduced in v1.21 and you don't use anything introduced in v1.22+).

**Reply by DevasiaThomas** [reply]
@clux Yup, I see your point. In an ideal world this would work fine. But I'll try out @teozkr first recommendation and then use solution 2 incase there is something I cannot absolutely get without using it.

**Reply by DevasiaThomas** [reply]
> @DevasiaThomas You would generally want to specify the oldest K8s version that contains both versions. In this case, that would be `v1_21` (since `CronJob` v1 was introduced in v1.21 and you don't use anything introduced in v1.22+).

Awesome, that makes sense! Thanks again :)

**Comment by nightkr** [comment]
No, you can only enable one version feature at a time. If you enable the
newer version then you need to disable the older one.
